### PR TITLE
Use body caching

### PR DESCRIPTION
**Pull request by jwolski**
@Raynos

### CODE REVIEW

**Comment by Raynos** [comment]
lgtm.

**Comment by mranney** [comment]
What does body caching do?

File: lib/request-proxy.js [hunk 1]
```diff
@@ -59,7 +59,7 @@ proto.proxyReq = function proxyReq(opts) {
     var timeout = opts.timeout ?
         opts.timeout : ringpop.proxyReqTimeout;
 
-    body(req, onBody);
+    body(req, res, { cache: true }, onBody);
 
     function onBody(err, rawBody) {
```
Comment (mranney) [inline lib/request-proxy.js]: It would be faster and less GC if you moved onBody back to the prototype here.

**Comment by jwolski** [comment]
allows multiple callers of `body()` to parse the request. if the body of the message has previously been parsed, it'll use that copy.

File: lib/request-proxy.js (hunk 1 above)
Comment (jwolski) [inline lib/request-proxy.js]: yea, this is a hot path that'll need careful attn. ill fix it in a different PR.

**Comment by Raynos** [comment]
@mranney 

 - The http sharding logic needs to peak into the body.
 - Ringpop needs to parse the body to proxy
 - Sirvice needs to parse the body to do json validation and stuff.

because multiple seperate modules all need the body and its a single stream we are using body caching so that `body(req, res, { cache: true }, cb)` works if called multiple times.

Caveat it will only work if called sequentially, we dont enable parallel caching.

**Pull request merged**